File: javaQuarkusShip/src/main/java/pl/battleships/javaquarkusship/service/GameServiceImpl.java
package pl.battleships.javaquarkusship.service;

import pl.battleships.javaquarkusship.api.dto.GameDto;
import pl.battleships.javaquarkusship.api.dto.GameStatusDto;
import pl.battleships.javaquarkusship.api.dto.PositionDto;
import pl.battleships.javaquarkusship.api.dto.ShotStatusDto;
import pl.battleships.core.api.BattleshipGame;
import pl.battleships.core.api.HistoryProvider;
import pl.battleships.javaquarkusship.mapper.GameMapper;


import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;
import java.util.List;
import java.util.stream.Collectors;

@ApplicationScoped
public class GameServiceImpl implements GameService {

    @Inject
    HistoryProvider historyProvider;

    @Inject
    BattleshipGame battleshipGame;

    @Inject
    GameMapper mapper;

    @Override
    public void joinTheGame(GameDto game) {
        battleshipGame.start(game.getId(), game.getSize(), game.getFirstShotIsYours());
    }

    @Override
    public ShotStatusDto opponentShot(String gameId, PositionDto position) {
        return mapper.map(
                battleshipGame.opponentShot(gameId, mapper.map(position))
        );
    }

    @Override
    public List<PositionDto> getAllShots(String id) {
        return historyProvider.getAllShots(id).stream().map(mapper::map).collect(Collectors.toList());
    }

    @Override
    public GameStatusDto getGameStatus(String gameId) {
        return mapper.map(battleshipGame.getGameStatus(gameId));
    }
}

File: kotlinQuarkusShip/src/main/kotlin/pl/battleships/kotlinquarkusship/service/GameServiceImpl.kt
package pl.battleships.kotlinquarkusship.service

import pl.battleships.api.dto.GameDto
import pl.battleships.api.dto.GameStatusDto
import pl.battleships.api.dto.PositionDto
import pl.battleships.api.dto.ShotStatusDto
import pl.battleships.core.api.BattleshipGame
import pl.battleships.core.api.HistoryProvider
import pl.battleships.kotlinquarkusship.extension.toGameStatusDto
import pl.battleships.kotlinquarkusship.extension.toPosition
import pl.battleships.kotlinquarkusship.extension.toPositionDto
import pl.battleships.kotlinquarkusship.extension.toShotStatusDto
import javax.enterprise.context.ApplicationScoped

@ApplicationScoped
class GameServiceImpl(
    private val battleshipGame: BattleshipGame,
    private val historyProvider: HistoryProvider, // trailing comma <3
) : GameService {

    override fun joinTheGame(game: GameDto) {
        battleshipGame.start(game.id, game.propertySize, game.firstShotIsYours)
    }

    override fun opponentShot(gameId: String, position: PositionDto): ShotStatusDto {
        return battleshipGame.opponentShot(gameId, position.toPosition()).toShotStatusDto()
    }

    override fun getAllShots(id: String): List<PositionDto> {
        return historyProvider.getAllShots(id).map { it.toPositionDto() }
    }

    override fun getGameStatus(gameId: String): GameStatusDto {
        return battleshipGame.getGameStatus(gameId).toGameStatusDto()
    }
}